cas-server-support-saml-idp/src/main/java/org/jasig/cas/support/saml/web/idp/profile/AbstractSamlProfileHandlerController.java

```diff
@@ -0,0 +1,368 @@
+            validator.setRenew(authnRequest.isForceAuthn());
+            final String serviceUrl = constructServiceUrl(request, response, authnRequest);
+            logger.debug("Created service url for validation: [{}]", serviceUrl);
+            final Assertion assertion = validator.validate(ticket, serviceUrl);
+            logCasValidationAssertion(assertion);
+            if (!assertion.isValid()) {
+                throw new SamlException("CAS assertion received is invalid");
+            }
+            final SamlRegisteredService registeredService = verifySamlRegisteredService(authnRequest);
+            final SamlMetadataAdaptor adaptor = getSamlMetadataAdaptorForService(registeredService, authnRequest);
+
+            logger.debug("Preparing SAML response for [{}]", adaptor.getEntityId());
+            responseBuilder.build(authnRequest, request, response, assertion, registeredService, adaptor);
+            logger.info("Built the SAML response for [{}]", adaptor.getEntityId());
+
+        } finally {
+            storeAuthnRequest(request, null);
+        }
+    }
+
+
+
+    /**
+     * Gets registered service and verify.
+     *
+     * @param authnRequest the authn request
+     * @return the registered service and verify
+     */
+    protected SamlRegisteredService verifySamlRegisteredService(final AuthnRequest authnRequest) {
+        final String serviceId = authnRequest.getAssertionConsumerServiceURL();
+        logger.debug("Checking service access in CAS service registry for [{}]", serviceId);
+        final RegisteredService registeredService =
+                this.servicesManager.findServiceBy(this.webApplicationServiceFactory.createService(serviceId));
+        if (registeredService == null || !registeredService.getAccessStrategy().isServiceAccessAllowed()) {
+            logger.warn("[{}] is not found in the registry or service access is denied. Ensure service is registered in the CAS registry",
+                    serviceId);
+            throw new UnauthorizedServiceException(UnauthorizedServiceException.CODE_UNAUTHZ_SERVICE);
+        }
+
+        if (registeredService instanceof SamlRegisteredService) {
+            final SamlRegisteredService samlRegisteredService = (SamlRegisteredService) registeredService;
+            logger.debug("Located SAML service in the registry as [{}] with the metadata location of [{}]",
+                    samlRegisteredService.getServiceId(), samlRegisteredService.getMetadataLocation());
+            return samlRegisteredService;
+        }
+        logger.error("Service [{}] is found in registry but it is not defined as a SAML service", serviceId);
+        throw new UnauthorizedServiceException(UnauthorizedServiceException.CODE_UNAUTHZ_SERVICE);
+    }
+
+    /**
+     * Initialize callback service.
+     *
+     * @param callbackUrl the callback url
+     * @return the service
+     */
+    protected Service registerCallback(final String callbackUrl) {
```

Comment:
What's the callback for? Does it work like the OAuth server support as a facade which delegates to the CAS server for authentication?